driver-core/src/main/java/com/datastax/driver/core/GuavaCompatibility.java

```diff
@@ -0,0 +1,215 @@
+ * <p>
+ * Therefore we depend on Guava 19, which has both the deprecated and the new APIs, and detect the actual version at
+ * runtime in order to call the relevant methods.
+ * <p>
+ * This is a hack, and might not work with subsequent Guava releases; the real fix is to stop exposing Guava in our
+ * public API. We'll address that in version 4 of the driver.
+ */
+public abstract class GuavaCompatibility {
+
+    private static final Logger logger = LoggerFactory.getLogger(GuavaCompatibility.class);
+    private static final Pattern GA_VERSION_EXTRACTOR = Pattern.compile("(\\d+\\.\\d+\\.\\d+).*");
+
+    public static final GuavaCompatibility INSTANCE = selectImplementation();
+
+    public static void init() {
+        // dummy method to force initialization of the class
+    }
+
+    public abstract <V> ListenableFuture<V> withFallback(ListenableFuture<? extends V> input,
+                                                         AsyncFunction<Throwable, V> fallback);
+
+    public abstract <V> ListenableFuture<V> withFallback(ListenableFuture<? extends V> input,
+                                                         AsyncFunction<Throwable, V> fallback, Executor executor);
+
+    public abstract <I, O> ListenableFuture<O> transformAsync(ListenableFuture<I> input,
+                                                              AsyncFunction<? super I, ? extends O> function);
+
+    public abstract <I, O> ListenableFuture<O> transformAsync(ListenableFuture<I> input,
+                                                              AsyncFunction<? super I, ? extends O> function,
+                                                              Executor executor);
+
+    public abstract boolean isSupertypeOf(TypeToken<?> target, TypeToken<?> argument);
+
+    public abstract Executor sameThreadExecutor();
+
+    private static GuavaCompatibility selectImplementation() {
+        String fullVersion = getBundleVersion(loadGuavaManifest());
+        // Get rid of potential rc qualifier, as it could throw off the lexical comparisons
+        String version = stripQualifiers(fullVersion);
+        if (version.compareTo("16.0.1") < 0) {
+            throw new DriverInternalError(String.format(
+                    "Detected incompatible version of Guava in the classpath (%s). " +
+                            "You need 16.0.1 or higher.", fullVersion));
+        } else if (version.compareTo("19.0") < 0) {
+            logger.info("Detected Guava {} in the classpath, using pre-19 compatibility layer", fullVersion);
+            return new Version18OrLower();
+        } else {
+            logger.info("Detected Guava {} in the classpath, using 19+ compatibility layer", fullVersion);
+            return new Version19OrHigher();
+        }
+    }
+
+    private static Manifest loadGuavaManifest() {
+        InputStream in = null;
+        try {
+            Enumeration<URL> resources = Resources.class.getClassLoader()
```

Comment:
I hadn't noticed it was beta. I think it's better to use a Guava class to make sure it's the classloader that loaded Guava, I've changed it to `Preconditions`.